File: internal/frontier/frontier.go
package frontier

import (
	"bytes"
	"compress/gzip"
	"crypto/md5"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"time"

	"github.com/TensorFu/larkdesk/internal/httpx"
	"github.com/TensorFu/larkdesk/internal/pb"
	"github.com/TensorFu/larkdesk/internal/session"
	"github.com/gorilla/websocket"
)

const (
	AppKeySaas     = "5f45da0e6c7a17dcba80494ef0ab9b21"
	AccessKeySalt  = "f8a69f1719916z"
	PassportTicket = "https://passport.feishu.cn/suite/passport/frontier_ticket/"
	FrontierHost   = "msg-frontier.feishu.cn"
)

func MaybeGunzip(data []byte) []byte {
	if len(data) < 2 || data[0] != 0x1f || data[1] != 0x8b {
		return data
	}
	r, err := gzip.NewReader(bytes.NewReader(data))
	if err != nil {
		return data
	}
	defer r.Close()
	out, err := io.ReadAll(r)
	if err != nil {
		return data
	}
	return out
}

func AccessKey(deviceID string) string {
	sum := md5.Sum([]byte("2" + AppKeySaas + deviceID + AccessKeySalt))
	return hex.EncodeToString(sum[:])
}

func FetchTicket(auth session.Auth) (deviceID, ticket string, err error) {
	if !auth.HasSessionCookie() {
		return "", "", fmt.Errorf("no decrypted session cookie for frontier")
	}
	u := PassportTicket + "?" + url.Values{"local_device_id": {""}}.Encode()
	req, err := http.NewRequest(http.MethodGet, u, nil)
	if err != nil {
		return "", "", err
	}
	req.Header.Set("User-Agent", "larkdesk/0.1")
	req.Header.Set("Cookie", auth.CookieHeader())
	req.Header.Set("Referer", "https://www.feishu.cn/messenger/")
	req.Header.Set("Origin", "https://www.feishu.cn")
	resp, err := httpx.Client(20 * time.Second).Do(req)
	if err != nil {
		return "", "", fmt.Errorf("frontier ticket %v", err)
	}
	defer resp.Body.Close()
	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", "", err
	}
	if resp.StatusCode >= 400 {
		return "", "", fmt.Errorf("frontier ticket HTTP %d", resp.StatusCode)
	}
	var body map[string]any
	if json.Unmarshal(raw, &body) != nil {
		return "", "", fmt.Errorf("frontier ticket was not JSON")
	}
	deviceID = fmt.Sprint(first(body, "device_id", "deviceId"))
	if deviceID == "<nil>" {
		deviceID = ""
	}
	ticket = fmt.Sprint(first(body, "ticket"))
	if ticket == "<nil>" {
		ticket = ""
	}
	if ticket == "" {
		return "", "", fmt.Errorf("frontier ticket missing")
	}
	if deviceID == "" {
		deviceID = "0"
	}
	return deviceID, ticket, nil
}

func first(m map[string]any, keys ...string) any {
	for _, k := range keys {
		if v, ok := m[k]; ok && v != nil {
			return v
		}
	}
	return ""
}

func BuildWSURL(deviceID, ticket string) string {
	q := url.Values{
		"access_key":      {AccessKey(deviceID)},
		"aid":             {"1"},
		"ticket":          {ticket},
		"device_id":       {deviceID},
		"fpid":            {"2"},
		"accept_encoding": {"gzip"},
		"request_id":      {pb.NewUUID()},
	}
	return "wss://" + FrontierHost + "/ws/v2?" + q.Encode()
}

func DecodeHeaders(raw []byte) map[string]string {
	headers := map[string]string{}
	for _, f := range pb.DecodeFields(raw) {
		if f.Num != 5 || f.Wire != "bytes" {
			continue
		}
		name, val := "", ""
		for _, inner := range pb.DecodeFields(f.Bytes) {
			if inner.Wire != "bytes" {
				continue
			}
			if inner.Num == 1 {
				name = string(inner.Bytes)
			} else if inner.Num == 2 {
				val = string(inner.Bytes)
			}
		}
		if name != "" {
			headers[name] = val
		}
	}
	return headers
}

func DecodeBody(raw []byte) []byte {
	var body []byte
	for _, f := range pb.DecodeFields(raw) {
		if f.Num == 8 && f.Wire == "bytes" {
			body = f.Bytes
		}
	}
	return MaybeGunzip(body)
}

type Frame struct {
	Headers map[string]string
	pb.Packet
}

func DecodeFrame(raw []byte) Frame {
	raw = MaybeGunzip(raw)
	headers := DecodeHeaders(raw)
	body := DecodeBody(raw)
	out := Frame{Headers: headers}
	if len(body) > 0 {
		enc := headers["content-encoding"]
		if enc == "" {
			enc = headers["Content-Encoding"]
		}
		if enc == "gzip" {
			body = MaybeGunzip(body)
		}
		out.Packet = pb.DecodePacket(body)
	}
	return out
}

func encodeHeader(k, v string) []byte {
	return pb.EncodeBytes(5, append(pb.EncodeString(1, k), pb.EncodeString(2, v)...))
}

func EncodeFrame(seq, ts uint64, packet []byte, userID string, cmd uint64) []byte {
	var b []byte
	b = append(b, pb.EncodeInt(1, seq)...)
	b = append(b, pb.EncodeInt(2, ts)...)
	b = append(b, pb.EncodeInt(3, 1)...)
	b = append(b, pb.EncodeInt(4, 1)...)
	cmdStr := fmt.Sprintf("%d", cmd)
	for _, kv := range [][2]string{
		{"content-type", "application/x-protobuf"},
		{"x-command", cmdStr},
		{"x-command-version", "2.7.0"},
		{"x-Source", "web"},
		{"X-Auth-User", userID},
		{"To-Cluster", "im"},
	} {
		b = append(b, encodeHeader(kv[0], kv[1])...)
	}
	b = append(b, pb.EncodeBytes(6, nil)...)
	b = append(b, pb.EncodeBytes(7, nil)...)
	b = append(b, pb.EncodeBytes(8, packet)...)
	return b
}

func Call(auth session.Auth, cmd uint64, payload []byte, echo []byte) ([]byte, error) {
	conn, err := Dial(auth)
	if err != nil {
		return nil, err
	}
	defer conn.Close()
	packet, _ := pb.EncodePacket(cmd, payload, "")
	frame := EncodeFrame(1, uint64(time.Now().UnixNano()), packet, auth.Identity.UserID, cmd)
	if err := conn.WriteMessage(websocket.BinaryMessage, frame); err != nil {
		return nil, err
	}
	_ = conn.SetReadDeadline(time.Now().Add(8 * time.Second))
	for i := 0; i < 16; i++ {
		_, raw, err := conn.ReadMessage()
		if err != nil {
			return nil, err
		}
		raw = MaybeGunzip(raw)
		if echo != nil && bytes.Contains(raw, echo) {
			return raw, nil
		}
		fr := DecodeFrame(raw)
		if fr.Cmd == cmd && len(fr.Payload) > 0 {
			return raw, nil
		}
	}
	return nil, fmt.Errorf("frontier call timeout cmd=%d", cmd)
}

func Dial(auth session.Auth) (*websocket.Conn, error) {
	deviceID, ticket, err := FetchTicket(auth)
	if err != nil {
		return nil, err
	}
	u := BuildWSURL(deviceID, ticket)
	d := websocket.Dialer{
		Proxy:             http.ProxyURL(httpx.ProxyURL()),
		HandshakeTimeout:  20 * time.Second,
		EnableCompression: true,
	}
	h := http.Header{}
	h.Set("Origin", "https://www.feishu.cn")
	h.Set("Cookie", auth.CookieHeader())
	h.Set("User-Agent", "larkdesk/0.1")
	conn, resp, err := d.Dial(u, h)
	if resp != nil && resp.Body != nil {
		resp.Body.Close()
	}
	if err != nil {
		return nil, err
	}
	conn.SetReadLimit(8 << 20)
	return conn, nil
}

File: src/larkdesk/frontier.py
"""Frontier WebSocket used by the Feishu web/desktop client for inbound push."""

from __future__ import annotations

import gzip
import hashlib
import json
import ssl
import uuid
from typing import AsyncIterator
from urllib.error import HTTPError, URLError
from urllib.parse import urlencode
from urllib.request import Request, build_opener

from larkdesk.gateway import _proxy_handler, decode_fields, decode_packet
from larkdesk.session import AuthMaterial

APPKEY_SAAS = "5f45da0e6c7a17dcba80494ef0ab9b21"
ACCESS_KEY_SALT = "f8a69f1719916z"
PASSPORT_TICKET = "https://passport.feishu.cn/suite/passport/frontier_ticket/"
FRONTIER_HOST = "msg-frontier.feishu.cn"


class FrontierError(RuntimeError):
    """Frontier ticket or websocket failed."""


def maybe_gunzip(data: bytes) -> bytes:
    if len(data) >= 2 and data[:2] == b"\x1f\x8b":
        try:
            return gzip.decompress(data)
        except OSError:
            return data
    return data


def access_key(device_id: str, *, appkey: str = APPKEY_SAAS) -> str:
    raw = f"2{appkey}{device_id}{ACCESS_KEY_SALT}".encode("utf-8")
    return hashlib.md5(raw).hexdigest()


def fetch_ticket(auth: AuthMaterial) -> tuple[str, str]:
    header = auth.cookie_header()
    if "session=" not in header and "osession=" not in header and "sl_session=" not in header:
        raise FrontierError("no decrypted session cookie for frontier")
    url = PASSPORT_TICKET + "?" + urlencode({"local_device_id": ""})
    req = Request(
        url,
        headers={
            "User-Agent": "larkdesk/0.1",
            "Cookie": header,
            "Referer": "https://www.feishu.cn/messenger/",
            "Origin": "https://www.feishu.cn",
        },
        method="GET",
    )
    handlers = []
    proxy = _proxy_handler()
    if proxy is not None:
        handlers.append(proxy)
    opener = build_opener(*handlers)
    try:
        with opener.open(req, timeout=20) as resp:
            body = json.loads(resp.read().decode("utf-8", "replace"))
    except HTTPError as exc:
        raise FrontierError(f"frontier ticket HTTP {exc.code}") from exc
    except URLError as exc:
        raise FrontierError(f"frontier ticket {exc.reason}") from exc
    except json.JSONDecodeError as exc:
        raise FrontierError("frontier ticket was not JSON") from exc
    if not isinstance(body, dict):
        raise FrontierError("frontier ticket unexpected body")
    device_id = str(body.get("device_id") or body.get("deviceId") or "")
    ticket = str(body.get("ticket") or "")
    if not ticket:
        raise FrontierError("frontier ticket missing")
    return device_id or "0", ticket


def build_ws_url(device_id: str, ticket: str) -> str:
    qs = urlencode(
        {
            "access_key": access_key(device_id),
            "aid": "1",
            "ticket": ticket,
            "device_id": device_id,
            "fpid": "2",
            "accept_encoding": "gzip",
            "request_id": str(uuid.uuid4()),
        }
    )
    return f"wss://{FRONTIER_HOST}/ws/v2?{qs}"


def decode_frontier_headers(raw: bytes) -> dict[str, str]:
    headers: dict[str, str] = {}
    for field, kind, value in decode_fields(raw):
        if field != 5 or kind != "bytes" or not isinstance(value, (bytes, bytearray)):
            continue
        name = ""
        val = ""
        for inf, ink, inv in decode_fields(bytes(value)):
            if inf == 1 and ink == "bytes" and isinstance(inv, (bytes, bytearray)):
                name = bytes(inv).decode("utf-8", "replace")
            elif inf == 2 and ink == "bytes" and isinstance(inv, (bytes, bytearray)):
                val = bytes(inv).decode("utf-8", "replace")
        if name:
            headers[name] = val
    return headers


def decode_frontier_body(raw: bytes) -> bytes:
    body = b""
    for field, kind, value in decode_fields(raw):
        if field == 8 and kind == "bytes" and isinstance(value, (bytes, bytearray)):
            body = bytes(value)
    return maybe_gunzip(body)


def decode_frontier_packet(raw: bytes) -> dict[str, object]:
    """Decode one websocket frame into an improto.Packet dict (cmd/payload/headers)."""
    raw = maybe_gunzip(raw)
    headers = decode_frontier_headers(raw)
    body = decode_frontier_body(raw)
    pkt: dict[str, object] = {"headers": headers}
    if body:
        encoding = headers.get("content-encoding") or headers.get("Content-Encoding") or ""
        if encoding == "gzip":
            body = maybe_gunzip(body)
        inner = decode_packet(body)
        pkt.update(inner)
    return pkt


async def iter_frontier_frames(auth: AuthMaterial) -> AsyncIterator[bytes]:
    """Yield raw websocket frames, reconnecting on drop."""
    import asyncio

    import websockets

    ssl_ctx = ssl.create_default_context()
    while True:
        try:
            device_id, ticket = fetch_ticket(auth)
            url = build_ws_url(device_id, ticket)
            async with websockets.connect(
                url,
                ssl=ssl_ctx,
                max_size=8 * 1024 * 1024,
                origin="https://www.feishu.cn",
                extra_headers={"Cookie": auth.cookie_header()},
                user_agent_header="larkdesk/0.1",
            ) as ws:
                async for msg in ws:
                    raw = msg if isinstance(msg, bytes) else str(msg).encode("utf-8", "replace")
                    yield maybe_gunzip(raw)
        except asyncio.CancelledError:
            raise
        except Exception:
            await asyncio.sleep(3)
